server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/Method.scala

```diff
@@ -33,6 +37,39 @@ trait Method {
 
   val requiredCapabilities: Capabilities
 
-  def process(capabilities: Set[CapabilityIdentifier], invocation: Invocation, mailboxSession: MailboxSession, processingContext: ProcessingContext): Publisher[(Invocation, ProcessingContext)]
+  def process(capabilities: Set[CapabilityIdentifier], invocation: Invocation, mailboxSession: MailboxSession, processingContext: ProcessingContext, sessionSupplier: SessionSupplier): Publisher[(Invocation, ProcessingContext)]
 }
 
+trait WithAccountId {
+  def accountId: AccountId
+}
+trait MethodRequiringAccountId[REQUEST <: WithAccountId] extends Method {
+  def metricFactory: MetricFactory
+
+  override def process(capabilities: Set[CapabilityIdentifier], invocation: Invocation, mailboxSession: MailboxSession, processingContext: ProcessingContext, sessionSupplier: SessionSupplier): Publisher[(Invocation, ProcessingContext)] = {
+    val result = (for {
+      request <- getRequest(mailboxSession, invocation)
+      response <- validateAccountId(request.accountId, mailboxSession, sessionSupplier, invocation).flatMap{
+        case Right(_) =>  doProcess(capabilities, invocation, mailboxSession, processingContext, sessionSupplier, request)
+        case Left(invocation) => SMono.just((invocation, processingContext))
+      }
+    } yield response)
+      .onErrorResume {
+      case e: IllegalArgumentException => SMono.just((Invocation.error(ErrorCode.InvalidArguments, e.getMessage, invocation.methodCallId), processingContext))
+      case e: Throwable => SMono.raiseError(e)
+    }
+
+    metricFactory.decoratePublisherWithTimerMetricLogP99(JMAP_RFC8621_PREFIX + methodName.value, result)
+  }
+
+  private def validateAccountId(accountId: AccountId, mailboxSession: MailboxSession, sessionSupplier: SessionSupplier, invocation: Invocation): SMono[Either[Invocation, Session]] = {
+    sessionSupplier.generate(mailboxSession.getUser)
+      .filter(session => session.accounts.map(_.accountId).contains(accountId))
+      .map(session => Right[Invocation, Session](session).asInstanceOf[Either[Invocation, Session]])
+      .switchIfEmpty(SMono.just(Left[Invocation, Session](Invocation.error(ErrorCode.AccountNotFound, invocation.methodCallId))))
+  }
+
+  def doProcess(capabilities: Set[CapabilityIdentifier], invocation: Invocation, mailboxSession: MailboxSession, processingContext: ProcessingContext, sessionSupplier: SessionSupplier, request: REQUEST): SMono[(Invocation, ProcessingContext)]
```

Comment:
Too much arguments to me. Would it makes sens to have case clssses to start grouping them a bit?

It could help simplifying method calls of individual methods as well...